### ISSUE TITLE
🌸 Cherry-pick request for #31230 into #31228 (Pending)

### ISSUE THREAD

**Issue by samouri** [ampproject/amphtml#31232]
# Cherry-pick request

|       Issue       |       PR       | Beta / Experimental? |   Stable?    |     LTS?     | [Release issue](https://github.com/ampproject/amphtml/labels/Type%3A%20Release) |
| :---------------: | :------------: | :------------------: | :----------: | :----------: | ------------------------------------------------------------------------------- |
| #31230 | #31230 |     **YES**     | **NO** | **NO** | #31228                                                            |

## Why does this issue meet the [cherry-pick criteria](https://github.com/ampproject/amphtml/blob/master/contributing/contributing-code.md#Cherry-picks)?

As a followup to an earlier cherrypick (https://github.com/ampproject/amphtml/issues/31004). That CP fixed a couple of issues, but there are still a few critical bugs leftover that are affecting amp4email, as well as newly found one. All of them were confirmed to be fixed by turning off the experiment.

These bugs are easily reproducible today on Android Gmail in production (e.g. by trying to view a dynamic mail with an amp-list.). By CPing we can resolve the user facing bugs ~1 week sooner.

Example bugs:
- amp-img with margin not to layout: `b/170491750`
- amp-list never to have its `layoutCallback()` fire: `b/170538870`, `b/169596105`

## List the steps to manually verify the changes in this cherry-pick

Reproing/verifying requires swapping out the active RTV within gmail's AMP on Android, which AFAIK is only possible via Android emulator. There are quite a few steps to get it all working.

The Dynamic Mail team confirmed that this PR fixes all of their bugs.

# Mini-postmortem

> **TODO:** This postmortem will be written after the cherry-pick deployment and before this issue is closed. Delete this TODO when the postmortem is ready.

## Impact

- amp-list doesn't fire in android gmail.

## Action Items

TODO

---

/cc @ampproject/release-on-duty @ampproject/wg-approvers @ampproject/cherry-pick-approvers

**Comment by rcebulko** [comment]
@samouri I'll start the process of getting a cherry-pick build going ASAP while we wait for approval. Matvei is out today, so once there is a cherry-pick RTV serving this, would you able to verify the fix the "final" beta/experimental RTV?

**Comment by samouri** [comment]
@rcebulko: sure thing!

**Comment by kristoferbaxter** [comment]
Approved. Thanks for moving quickly.

**Comment by samouri** [comment]
@rcebulko: fix has been verified by the Dynamic Mail team, using `2011140244002`

**Comment by rsimha** [comment]
_[Cleaning up old CP issues]_ Can this be closed?

**Comment by samouri** [comment]
I believe this was already merged in, so yes.

**Comment by rsimha** [comment]
> I believe this was already merged in, so yes. Is the responsibility usually on the CP opener or the release eng to close the issue?

Typically, the CP opener is expected to fill out the post mortem section after the CP has been merged.

**Comment by samouri** [comment]
**Mini post-mortem**
There were 3 separate bugs in the `intersect-resources` experiment which only affected amp4email. 
- One was due to a typo, using `this.relayoutAll_` instead of `relayoutAll` (https://github.com/ampproject/amphtml/pull/31200)
- A second was due to a race condition where the size of an element changes in between  InOb's measurement capture and application caused by a change to document size (https://github.com/ampproject/amphtml/pull/30937).
- And the third was due to a race condition w.r.t to viewport services `resize` events (https://github.com/ampproject/amphtml/pull/31376). This was only reproducible in a viewer environment where all of the assets are cached (html/css/js).

**impact**
- The most heavily impacted surface was Android gmail. Certain emails wouldn't load correctly until ~30s in.
- There were also issues with `<amp-list>` in docs digest emails getting cutoff until a pass was forced.

**action items**
We should create e2e viewer tests.

One of the factors that made these bugs so tricky to detect + reproduce is that we weren't sure what set of conditions were required to reproduce. This past week I researched/learned how to reproduce with a Viewer. We essentially need to:
- Test various initial `documentHeight`'s for the viewer iframe (0px, 1px, 100px)
- The iframe's height should be dynamically updated based on `documentHeight` messages from the viewer
- We need to test permutations where all of the HTML/CSS/JS assets are cached. This creates unique race conditions.

Once we have this suite of e2e tests running, I'll be significantly more confident that amp4email is stable.
